cmd/wait/runner.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	{
+		r.logger.LogCtx(ctx, "message", "waiting for cluster API to be reachable")
+
+		o := func() error {
+			_, err := k8sClient.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+			if tenant.IsAPINotAvailable(err) {
+				r.logger.LogCtx(ctx, "message", "API not yet available")
+				return microerror.Mask(err)
+			} else if err != nil {
+				r.logger.LogCtx(ctx, "message", "unexpected error contacting API", "error", err)
+				return microerror.Mask(err)
+			}
+			return nil
+		}
+
+		b := backoff.NewMaxRetries(60, 20*time.Second)
+
+		err = backoff.Retry(o, b)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "message", "API is now reachable")
+	}
+
+	{
+		r.logger.LogCtx(ctx, "message", "waiting for nodes to be ready")
+
+		o := func() error {
+			nodes, err := k8sClient.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+			if err != nil {
+				r.logger.LogCtx(ctx, "message", "error listing nodes", "error", err)
+				return microerror.Mask(err)
+			}
+			nodeCount := len(nodes.Items)
+			readyCount := 0
+			for _, node := range nodes.Items {
+				for _, condition := range node.Status.Conditions {
+					if condition.Type == "Ready" {
+						if condition.Status == "True" {
+							readyCount++
+						}
+						break
+					}
+				}
+			}
+			if nodeCount < 2 {
+				message := fmt.Sprintf("found %d registered nodes, waiting for at least 2", nodeCount)
+				r.logger.LogCtx(ctx, "message", message)
+				return microerror.Mask(errors.New(message))
+			}
+			if readyCount < nodeCount {
+				message := fmt.Sprintf("%d out of %d nodes ready", readyCount, nodeCount)
+				r.logger.LogCtx(ctx, "message", message)
+				return microerror.Mask(errors.New(message))
+			}
+			return nil
+		}
+
+		b := backoff.NewMaxRetries(60, 20*time.Second)
+
+		err = backoff.Retry(o, b)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		r.logger.LogCtx(ctx, "message", "nodes are ready")
+	}
+
+	{
+		r.logger.LogCtx(ctx, "message", "waiting for CoreDNS to be ready")
+
+		targetLabels := map[string]string{
+			"kubernetes.io/cluster-service": "true",
+			"kubernetes.io/name":            "CoreDNS",
+		}
+
+		o := func() error {
+			serviceLabelSelector := labelsToSelector(targetLabels)
+			services, err := k8sClient.CoreV1().Services("kube-system").List(ctx, v1.ListOptions{
+				LabelSelector: serviceLabelSelector,
+			})
+			if err != nil {
+				r.logger.LogCtx(ctx, "message", "error listing services", "error", err)
+				return microerror.Mask(err)
+			}
+			if len(services.Items) == 0 {
+				message := fmt.Sprintf("CoreDNS service not found using label selector %#q", serviceLabelSelector)
+				r.logger.LogCtx(ctx, "message", message)
+				return microerror.Mask(errors.New(message))
+			}
+
+			service := services.Items[0]
+			podLabelSelector := labelsToSelector(service.Spec.Selector)
+			pods, err := k8sClient.CoreV1().Pods("kube-system").List(ctx, v1.ListOptions{
```

Comment:
This is the best method I could come up with for ensuring that CoreDNS is actually running and healthy since we could have a service with no running pods.

Reply:
IMO the pods being ready is a pretty solid start, but just to double check, would this also be happy with completed init containers?

Reply:
There are two other fields for init containers (`.Status.InitContainerStatuses`) and ephemeral containers (`.Status.EphemeralContainerStatuses`) so I think this will just tell us about the main pod containers.